### ISSUE TITLE
Wrong tracked region for ValueObservation?

### ISSUE THREAD

**Issue by rpoelstra** [groue/GRDB.swift#861]
Hi,

I've a `ValueObservation` that performs the following query:
```
// Fetch layers
let layers = try Layer.fetchAll(db)

// Fetch related images
let layerIds = layers.map(\.id!)
let alias=TableAlias()
let imageFiles = try ImageFile
    .filter(layerIds.contains(ImageFile.Columns.layerId))
    .order(ImageFile.Columns.url)
    .annotated(with: ImageFile.errors.count)
    .asRequest(of: ImageWithErrorCount.self)
    .fetchAll(db)
```

This observation doesn't properly track changes to the `imageFile` table. The tracked region is reported as:
`imageFileError(id,imageId),layer(<bunch of colums>)`, which does indeed not list the `imageFile` table.

Is this a bug?

Thanks in advance.

Kind regards,

Remco Poelstra

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.0.1
**Installation method:** manual
**Xcode version:** 11.7
**Swift version:** 5
**Platform(s) running GRDB:** macOS
**macOS version running Xcode:** 10.15.7

**Comment by groue** [comment]
Hello @rpoelstra,

> Is this a bug?

I don't know.

Would you please setup a reproducible case?

```swift
// Records
struct Layer { ... }
struct ImageFile { ... }

// Database schema
var config = Configuration()
config.prepareDatabase { db in
    db.trace { print($0) }
}
let dbQueue = DatabaseQueue(configuration: config)
try dbQueue.write { db in /* create tables */ }

// Observation
let observation = ValueObservation
    .tracking { db in ... }
    .print()
_ = observation.start(
    in: dbQueue,
    scheduling: .immediate,
    onError: { error in },
    onChange: { _ in })
 ```

**Comment by rpoelstra** [comment]
Hi,

Please find attached a simpel test project.

Remco
[GRDBValueObserver.zip](https://github.com/groue/GRDB.swift/files/5457742/GRDBValueObserver.zip)

**Comment by groue** [comment]
Thank you @rpoelstra. Congrats, you found a bug! It is triggered when there are no layers in the database.

A fix will ship shortly. Meanwhile, add a dummy request to `imageFile` in your observation. For example:

```swift
let observation = ValueObservation.tracking { db -> [LayerImagesListItem] in
    ...

    #warning("TODO: remove when https://github.com/groue/GRDB.swift/issues/861 is fixed")
    _ = try ImageFile.fetchCount(db)
    
    return items
}

```

**Comment by rpoelstra** [comment]
Hi,

Did you mean when there are no ImageFileError's in the database? As the bug is clearly triggered with layers in the database, as the test project shows.

**Comment by rpoelstra** [comment]
Ah, wait. When the observation is created there are no layers, of course.

**Comment by groue** [comment]
> Ah, wait. When the observation is created there are no layers, of course.

Yes, exactly.

**Comment by groue** [comment]
After investigation:

This issue reveals a surprising behavior. I agree something has to be done in order to prevent it.

In the same time, it's not a bug. Actually everything is behaving as expected.

In the current state of GRDB, the proper fix is not to add a dummy request to `ImageFile`. It is to replace `ValueObservation.tracking { }` with `ValueObservation.trackingVaryingRegion { }`. See [Observing a Varying Database Region](https://github.com/groue/GRDB.swift#observing-a-varying-database-region).

Let me try to explain:

When there is no layer, `ImageFile.filter(layerIds.contains(ImageFile.Columns.layerId))` is a empty request. GRDB sees that, and does not add the `imageFile` table to the tracked region: there is no point tracking an empty request.

The goal is to help observations of individual rows. When it is possible to know which [rowid](https://sqlite.org/lang_createtable.html#rowid) of a particular table are fetched by a request, an observation is able to ignore modifications to other rows (this is a useful optimization).

```swift
// Tracks row 1. Modifications to other rows do not trigger the observation.
let obs = ValueObservation.tracking { db in
    try Player.filter(key: 1).fetchOne(db)
}

// Tracks rows 1, 2, 3. Modifications to other rows do not trigger the observation.
let obs = ValueObservation.tracking { db in
    try Player.filter(keys: [1, 2, 3]).fetchAll(db)
}

// Tracks nothing. Nothing triggers the observation.
let obs = ValueObservation.tracking { db in
    try Player.none().fetchAll(db)
}

// Tracks nothing. Nothing triggers the observation.
let emails: [String] = [] // empty array
let obs = ValueObservation.tracking { db in
    try Player.filter(emails.contains(Player.Columns.email).fetchAll(db)
}
```

Rowids, when it is possible to infer them from the request, are *part of the tracked region*.

Things turn more subtle when rowids depend on the results of other requests:

```swift
// HERE BE DRAGONS
// Always tracks the same player
let obs = ValueObservation.tracking { db -> Player? in
    let playerId = try Preferences.fetchOne(db)?.favoritePlayerId
    try Player.filter(key: playerId).fetchOne(db)
}

// HERE BE DRAGONS
// May track no player at all if there is no team on the first fetch.
let obs = ValueObservation.tracking { db -> Player? in
    let teamIDs = try Team.select(Team.Columns.id, as: Int64.self).fetchAll(db)
    try Player.filter(teamIDs.contains(Player.Columns.teamId).fetchAll(db)
}
```

Those above observations really are [observations of a varying database region](https://github.com/groue/GRDB.swift#observing-a-varying-database-region), due to their vaying sets of rowids.

The difficult question asked by this issue is: **how is it possible to preserve the optimization of the first observations, while disabling it for the "dragon" observations?**

So let me please repeat myself. In the current state of GRDB, the proper fix is to replace `ValueObservation.tracking { }` with `ValueObservation.trackingVaryingRegion { }`, because your app is observing a varying region.

Maybe we'll solve this surprise by changing the behavior of the library (any suggestion?)

Maybe we'll preserve the behavior, but enhance the documentation so that this topic is made much more clear?

**Comment by groue** [comment]
> Maybe we'll solve this surprise by changing the behavior of the library (any suggestion?)

Note that if we would not optimize the region for **empty** requests, we'd "support" your use case:

```swift
// If there is no team on the first fetch, then we notice that
// we don't fetch any player, but we keep on observing the
// player table anyway, because we're nice.
let obs = ValueObservation.tracking { db -> Player? in
    let teamIDs = try Team.select(Team.Columns.id, as: Int64.self).fetchAll(db)
    try Player.filter(teamIDs.contains(Player.Columns.teamId).fetchAll(db)
}
```

But we have to think of other users, and find the good balance between consistency and efficiency. If we'd provide a special tracking for empty requests, then:

```swift
// NOT CONSISTENT
// Would still track the same player, regardless
// of the modifications to the favorite player
let obs = ValueObservation.tracking { db -> Player? in
    let playerId = try Preferences.fetchOne(db)?.favoritePlayerId
    try Player.filter(key: playerId).fetchOne(db)
}

// NOT EFFICIENT
// Would track the player table for no good reason.
let obs = ValueObservation.tracking { db in
    try Player.none().fetchAll(db)
}
```

**Comment by rpoelstra** [comment]
This is an interesting side-effect indeed.

I've no problems with using `ValueObservation.trackingVaryingRegion { }`, but I think in this particular situation I will just fetch all images and only use those that actually belong to a layer. This may well be faster as very few images do not belong to a layer.

It would indeed be a good idea to update the documentation to note that the row id's are part of the tracked region 

And maybe as well indicate which optimisations are not performed for `ValueObservation.trackingVaryingRegion { }` as the documentation only states that some are prevented. This would allow someone (like me) to judge whether it's better to fetch too many rows or use the varying region tracking.

Thanks!

**Comment by groue** [comment]
> And maybe as well indicate which optimisations are not performed for ValueObservation.trackingVaryingRegion { } as the documentation only states that some are prevented. This would allow someone (like me) to judge whether it's better to fetch too many rows or use the varying region tracking.

You're right, this needs to be explained. Observations of varying regions can not profit from DatabasePool's ability to perform concurrent readonly requests. They must fetch their fresh values from the writer database connection, and thus postpone other application components that want to write. Observations of varying regions increase database write contention.

**Comment by groue** [comment]
Closing in favor of #862

**Comment by groue** [comment]
#868, shipped in v5.2.0, has reversed the default tracking mode. Now `ValueObservation.tracking(_:)` won't miss a change, even with varying regions.
